pypeit/alignframe.py

```diff
@@ -49,295 +121,176 @@ class Alignment():
-            if show_trace:
-                self.show('overplot', chname='align_traces', align_traces=align_traces, slits=False)
-            align_prof['{0:d}'.format(sl)] = align_traces.copy()
-
-        align_dict = self.generate_dict(align_prof)
+                msgs.error("Alignment tracing has failed on slit {0:d}".format(slit_idx))
+            align_prof['{0:d}'.format(slit_idx)] = align_traces.copy()
 
         # Steps
         self.steps.append(inspect.stack()[0][3])
 
         # Return
-        return align_dict
+        return align_prof
 
-    def generate_dict(self, align_prof):
+    def generate_traces(self, align_prof):
         """
         Generate a dictionary containing all of the information from the profile fitting
 
-        Args:
-            align_prof (:obj:`dict`):
-                Dictionary of SpecObjs classes (one for each slit)
+        Parameters
+        ----------
+        align_prof : :obj:`dict`
+            Dictionary of SpecObjs classes (one for each slit)
+
+        Returns
+        -------
+        align_traces : :obj:`numpy.ndarray`
+            Spatial traces (3D array of shape [nspec, ntraces, nslits])
         """
-        nbars = len(self.par['locations'])
-        nspec, nslits = self.slit_left.shape
+        nbars = len(self.alignpar['locations'])
         # Generate an array containing the centroid of all bars
-        self._alignprof = np.zeros((nspec, nbars, nslits))
-        for sl in range(nslits):
+        align_traces = np.zeros((self.nspec, nbars, self.nslits))
+        for sl in range(self.nslits):
             sls = '{0:d}'.format(sl)
             for bar in range(nbars):
                 if align_prof[sls][bar].SLITID != sl:
-                    msgs.error("Alignment profiling failed to generate dictionary")
-                    self._alignprof[:, bar, sl] = align_prof[sls][bar].TRACE_SPAT
-        # Return the profile information as a single dictionary
-        return
-
-    def save(self, outfile, overwrite=True, checksum=True, float_dtype='float32',
-             master_dir=None, master_key=None):
-        """
-        Save the alignment traces to a master frame.
-
-        Args:
-            outfile (:obj:`str`):
-                Name for the output file.  Defaults to
-                :attr:`master_file_path`.
-            overwrite (:obj:`bool`, optional):
-                Overwrite any existing file.
-            checksum (:obj:`bool`, optional):
-                Passed to `astropy.io.fits.HDUList.writeto`_ to add
-                the DATASUM and CHECKSUM keywords fits header(s).
-            float_dtype (:obj:`str`, optional):
-                Convert floating-point data to this data type before
-                writing.  Default is 32-bit precision.
-        """
+                    msgs.error("Alignment profiling failed to generate traces")
+                align_traces[:, bar, sl] = align_prof[sls][bar].TRACE_SPAT
+        return align_traces
```

Comment:
May not be an issue, and I guess this goes back to the use of `extract.objfind` to get the traces, but the nested `for` loop here feels inefficient.  Is there a way to optimize this better?  We can do that in a different PR.

Reply:
This part is super fast compared to the object tracing, so I'm not too concerned. I think the problem is because the object tracing is stored in a dictionary. I'm not really sure how to convert this into a single for loop without changing the `extract.objfind`...